File: Fall 2023/MA 305/Labs/lab 0/lab0-files/lab0.f
!++++++++++++++++   compile with gfortran Kahan.f   ++++++++++++++++++++
!=======================================================================
!      W. Kahan's example of finite precision arithmetic failings 
!  from C. Van Loan, Intro. to Computational Science, 1995; p.xxiv     
!####&###1#########2#########3#########4#########5#########6#########7##
         program Kahan
         
         implicit none 
         real :: h, x, y, u, v, q
         !double precision :: h, x, y, u, v, q
         
         h=1.0/2
         x=2.0/3 - h 
         y=3.0/5 - h 
         u=(x+x+x) - h
         v=(y+y+y+y+y) - h
         q=u/v 

         write(*,*) 
         write(*,*)'Fortran:' 
         write(*,*) 'h =', h
         write(*,*) 'x =', x
         write(*,*) 'y =', y
         write(*,*) 'u =', u
         write(*,*) 'v =', v
         write(*,*) 'q =', q
         write(*,*) 

        stop

      end program kahan 
!_______________________________________________________________________
!####&###1#########2#########3#########4#########5#########6#########7##

File: Fall 2023/MA 305/(003) Labs/lab 0/lab0-files/lab0.c
/* 
++++++++++++++++   compile with gcc kahan.c     ++++++++++++++++++++++++
========================================================================
     W. Kahan's example of finite precision arithmetic failings 
  from C. Van Loan, Intro. to Computational Science, 1995; p.xxiv     
________________________________________________________________________
*/ 

#include <stdio.h>
#include <math.h>

  int main() {

  double h, x, y, u, v, q; 

 	h = 1.0/2;
 	x = 2.0/3 - h;
 	y = 3.0/5 - h;
	u = (x + x + x) - h;
	v = (y + y + y + y + y) - h;
	q = u/v;

  	printf("\nC:\n");
  	printf("h = %0.16f\n",h);
  	printf("x = %0.16f\n",x);
  	printf("y = %0.16f\n",y);
  	printf("u = %0.16f\n",u);
  	printf("v = %0.16f\n",v);
  	printf("q = %0.16f\n",q);
  	printf("\n");

}
